private/bufpkg/bufreflect/bufreflect.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufreflect
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufimage"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmoduleref"
+	"github.com/bufbuild/buf/private/pkg/app/appcmd"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// NewMessage returns a new dynamic proto.Message for the fully qualified typeName
+// in the bufimage.Image.
+func NewMessage(
+	ctx context.Context,
+	image bufimage.Image,
+	typeName string,
+) (proto.Message, error) {
+	if err := validateTypeName(typeName); err != nil {
+		return nil, err
+	}
+	files, err := protodesc.NewFiles(bufimage.ImageToFileDescriptorSet(image))
+	if err != nil {
+		return nil, err
+	}
+	descriptor, err := files.FindDescriptorByName(protoreflect.FullName(typeName))
+	if err != nil {
+		return nil, err
+	}
+	typedDescriptor, ok := descriptor.(protoreflect.MessageDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("%q must be a message but is a %T", typeName, descriptor)
+	}
+	return dynamicpb.NewMessage(typedDescriptor), nil
+}
+
+// ParseSourceAndType returns the moduleReference and typeName from the source and type provided by the user.
+// When source is not provided, we assume the type is a fully-qualified path to the type and try to parse it.
+// Otherwise, if both source and type are provided, the type must be a valid Protobuf identifier (e.g. weather.v1.Units).
+func ParseSourceAndType(
+	ctx context.Context,
+	flagSource string,
+	flagType string,
+) (moduleReference string, typeName string, _ error) {
```

Comment:
Given that this function is particularly tailored to flag parameters, it probably belongs in the `bufcli` package. We might be able to justify that for `parseFullyQualifiedPath` too, but we could export `validateTypeName` from here so it can be used by both `bufreflect.NewMessage` and the `bufcli` functions.